src/storage/mvcc/txn.rs

```diff
@@ -205,9 +219,103 @@ impl<S: Snapshot> MvccTxn<S> {
         Ok(())
     }
 
+    pub fn acquire_pessimistic_lock(
+        &mut self,
+        key: Key,
+        primary: &[u8],
+        for_update_ts: u64,
+        options: &Options,
+    ) -> Result<()> {
+        if let Some(lock) = self.reader.load_lock(&key)? {
+            if lock.ts != self.start_ts {
+                return Err(Error::KeyIsLocked {
+                    key: key.into_raw()?,
+                    primary: lock.primary,
+                    ts: lock.ts,
+                    ttl: lock.ttl,
+                });
+            }
+            if lock.lock_type != LockType::Pessimistic {
+                return Err(Error::LockTypeNotMatch {
+                    start_ts: self.start_ts,
+                    key: key.into_raw()?,
+                    pessimistic: false,
+                });
+            }
+            MVCC_DUPLICATE_CMD_COUNTER_VEC.pessimistic_lock.inc();
+            return Ok(());
+        }
+
+        if let Some((commit_ts, _)) = self.reader.seek_write(&key, u64::max_value())? {
+            // The isolation level of pessimistic transactions is RC. `for_update_ts` is
+            // the commit_ts of the data this transaction read. If exists a commit version
+            // whose commit timestamp is larger than current `for_update_ts`, the
+            // transaction should retry to get the latest data.
+            if commit_ts > for_update_ts {
+                MVCC_CONFLICT_COUNTER.pessimistic_lock_conflict.inc();
+                return Err(Error::WriteConflict {
+                    start_ts: self.start_ts,
+                    // TiDB will try to extract the latest `commit_ts` from WriteConflict
+                    // error to avoid getting a timestamp from tso. So we set
+                    // `conflict_start_ts` to `commit_ts` here.
+                    conflict_start_ts: commit_ts,
```

Comment:
Why? I still do not understand

Reply:
https://github.com/tikv/tikv/blob/master/src/server/service/kv.rs#L1704
https://github.com/pingcap/tidb/blob/master/executor/adapter.go#L433
When TiDB got `WriteConflict` in pessimistic transactions, it tries to extract the latest `commit_ts` as new `for_update_ts` to get the latest data, so avoid getting a timestamp from pd again. But in TiKV, we set `conflict_start_ts` as `conflict_ts`.

Reply:
Will add a `conflict_commit_ts` to kvproto.